lib/oli/delivery/sections/section_resource_depot.ex

```diff
@@ -0,0 +1,154 @@
+  directly.
+  """
+  def get_full_hierarchy(%Section{} = section) do
+    init_if_necessary(section.id)
+
+    page = Oli.Resources.ResourceType.id_for_page()
+    container = Oli.Resources.ResourceType.id_for_container()
+
+    srs = Depot.query(@depot_desc, section.id, [{:resource_type_id, {:in, [page, container]}}])
+    Oli.Delivery.Hierarchy.full_hierarchy(section, srs)
+  end
+
+  @doc """
+  Retrieve the full hierarchy of section resources for a given section as would
+  the DeliveryResolver.
+
+  Usefull for generating a full hierarchy of section resources for a section,
+  in places where code had been calling Oli.Publishing.DeliveryResolver.full_hierarchy
+  directly.
+  """
+  def get_delivery_resolver_full_hierarchy(%Section{} = section) do
+    init_if_necessary(section.id)
+
+    page = Oli.Resources.ResourceType.id_for_page()
+    container = Oli.Resources.ResourceType.id_for_container()
+
+    srs = Depot.query(@depot_desc, section.id, [{:resource_type_id, {:in, [page, container]}}])
+    Oli.Publishing.DeliveryResolver.full_hierarchy(section, srs)
+  end
+
+  @doc """
+  Returns a list of SectionResource records for all graded pages for a given section.
+  """
+  def graded_pages(section_id) do
+    init_if_necessary(section_id)
+
+    page = Oli.Resources.ResourceType.id_for_page()
+
+    Depot.query(@depot_desc, section_id, graded: true, resource_type_id: page)
+    |> Enum.sort_by(& &1.numbering_index)
+  end
+
+  @doc """
+  Access the SectionResource records pertaining to the course schedule.
+  """
+  def retrieve_schedule(section_id, filter_resource_type \\ false) do
+    init_if_necessary(section_id)
+
+    page_type_id = Oli.Resources.ResourceType.id_for_page()
+    container_type_id = Oli.Resources.ResourceType.id_for_container()
+
+    filter_by_resource_type =
+      case filter_resource_type do
+        :pages ->
+          [{:resource_type_id, {:==, page_type_id}}]
+
+        :containers ->
+          [{:resource_type_id, {:==, container_type_id}}]
+
+        _ ->
+          [{:resource_type_id, {:in, [container_type_id, page_type_id]}}]
+      end
+
+    Depot.query(@depot_desc, section_id, filter_by_resource_type)
+  end
+
+  @doc """
+  Returns a list of SectionResource pages (graded + ungraded) for a given section.
+  An optional parameter `graded_only` can be passed to filter only graded pages.
+  """
+  def get_lessons(section, graded_only \\ false) do
+    init_if_necessary(section.id)
+
+    page_type_id = Oli.Resources.ResourceType.id_for_page()
+
+    conditions =
+      case graded_only do
+        true -> [resource_type_id: page_type_id, graded: true, hidden: false]
+        false -> [resource_type_id: page_type_id, hidden: false]
+      end
+
+    Depot.query(@depot_desc, section.id, conditions)
+  end
+
+  defp init_if_necessary(section_id) do
+    if Depot.table_exists?(@depot_desc, section_id) do
+      {:ok, :exists}
+    else
+      if SectionResourceMigration.requires_migration?(section_id) do
```

Comment:
Why are we migrating data in runtime and not as part of a separate migration?

Reply:
I went down the path of a just in time migration to avoid a potentially large and expensive migration at deployment time.